PublishTestResultScreenshotV0/index.ts

```diff
@@ -37,26 +45,76 @@ async function run() {
+/**
+ * Given system [build or release] returning the Id of the latest TestRun 
+ * @param system 
+ * @returns Latest TestRun Id
+ */
+async function getTestRunIdBy(system : string): Promise<number> {
+    let ret :number = -1;
+    let testRuns : TestRun[] = [];
+    let today = new Date();
+    let yesterday = new Date();
+    yesterday.setDate(today.getDate() -1);
+
+    let buildid = tl.getVariable("BUILD_BUILDID");
+    switch(system) {
+        case "build":
+            testRuns = await testApi.queryTestRuns(project, yesterday, today, undefined, undefined, undefined, undefined, [+buildid], undefined, undefined, undefined, undefined, undefined, undefined,undefined, undefined,undefined );
+            break;
+        case "release":
+            let releaseId = tl.getVariable("RELEASE_RELEASEID");
+            let releaseEnvId = tl.getVariable("RELEASE_ENVIRONMENTID");
+            testRuns = await testApi.queryTestRuns(project, yesterday, today, undefined, undefined, undefined, undefined, [+buildid], undefined, undefined, [+releaseId], undefined, [+releaseEnvId], undefined,undefined, undefined,undefined );
+            break;
+        default:
+            tl.debug(`unsupported system value: ${system}`);
+    }
+
+    if (testRuns?.length === 1){
+        ret = testRuns[0].id;
+    }
+    else if (testRuns?.length > 1) {
+        // when query return more than 1 Run, sort desc and pick the first.
+        ret = testRuns.map( run=> run.id).sort((a,b)=> b-a)[0]; 
+    }
+
+    return ret;
+}
+
+/**
+ * Give a runId return an array of TestCaseResult which TestOutcome is failed
+ * @param runId 
+ * @returns an array of TestCaseResult
+ */
+async function getFailureResultsBy(runId: number) :  Promise<TestCaseResult[]>{
+    let ret : TestCaseResult[] = [];
+    ret = await testApi.getTestResults(project, runId, ResultDetails.None, undefined, undefined,[TestOutcome.Failed] );
+    return ret;
+}
+
+
+async function uploadScreenshots(failedTests: TestCaseResult[]) {
     let apiCalls: Promise<any>[] = [];
     let missingScreenshots: Error[] = [];
-    let totalFailures = failedTests.length
+    let totalFailures = failedTests?.length ?? 0;
 
     if(totalFailures <= 0) {
         tl.setResult(tl.TaskResult.Skipped, "No test failures found.")
         return
     }
     console.log(totalFailures + " tests failed. Will proceed with screenshot upload.")
-    failedTests.forEach(async (failedTest: ShallowTestCaseResult, index) => {
-        let testName = failedTest.automatedTestName;
+    failedTests.forEach(async (failedTest: TestCaseResult, index) => {
+        let testName = failedTest.automatedTestName ?? '';
         let className = failedTest.automatedTestStorage;
-        let imgPath = getScreenshotFolder() + className + "/" + testName + ".png";//TODO make it configurable in upcoming version
+        let imgPath = getScreenshotFolder() + className + "/" + filenamify(testName) + ".png";//TODO make it configurable in upcoming version
```

Comment:
I'm curious, why was that new `filenamify` dependency needed for you ? 
Any specific problems you were having with your screenshot file names you needed to add support for?

Reply:
Some test description contain non system path friendly character, like ":" . Using filenamify here, simply to convert those characters to be usable in this extension.